### ISSUE TITLE
Include `dts` file for `dist/client/theme-default/composables`

### ISSUE THREAD

**Issue by xsjcTony** [vuejs/vitepress#2595]
### Is your feature request related to a problem? Please describe.

Imagine the use case:

I want to build a custom page type based on `frontmatter` feature, e.g.
```yaml
---
layout: api
# requests:
#   - - method: get
#     - url: ...
---
```

But as I want to inherit all features of sidebars, content footer, etc., hence, I'll need to copy the whole `VPContent.vue` file too add one more custom component, and override internal `VPContent` component by alias, e.g.
```vue
<VPApi v-else-if="frontmatter.layout === 'api'">...</VPApi>
```

But when I'm trying to resolve those import paths, I'm running into an issue that the type of `composables` used is lost, because it's only `.js` file, and does not have any corresponding `dts` file to declare their types.

```ts
// from
import NotFound from '../NotFound.vue'
import { useData } from '../composables/data'
import { useSidebar } from '../composables/sidebar'
// ...other `.vue` components

// to
import NotFound from 'vitepress/dist/client/theme-default/NotFound.vue';
// This is fine, because it's exported from `vitepress` straight away and have type declarations
import { useData } from 'vitepress'; 
// But this leads into problems, as it has no type declarations
import { useSidebar } from 'vitepress/dist/client/theme-default/composables/sidebar.js';
// ...other `.vue` components
```

All types are then lost:
![Snipaste_2023-07-09_18-30-01](https://github.com/vuejs/vitepress/assets/45784210/e91c8e6d-f360-48dc-afa3-e1d3e95bb7d4)

### Describe the solution you'd like

Include all `dts` files when building vitepress, all provide any other way for users to import corresponding types.

Or alternatively, any better suggestion on building the features I want? Since I need to do a lot of customizations there, so instead of using `markdown`, I'd rather define my own components but still has the ability to benefit from built-in vitepress features

### Validations

- [X] Follow our [Code of Conduct](https://vuejs.org/about/coc.html)
- [X] Read the [docs](https://vitepress.dev).
- [X] Read the [Contributing Guidelines](https://github.com/vuejs/vitepress/blob/main/.github/contributing.md).
- [X] Check that there isn't already an issue that asks for the same feature to avoid creating a duplicate.

**Comment by brc-dd** [comment]
You don't need to override VPContent to add another layout. Something like this should work:

```vue
<!-- .vitepress/theme/Layout.vue -->

<script setup lang="ts">
import { useData } from 'vitepress'
import DefaultTheme from 'vitepress/theme'

const { frontmatter } = useData()
</script>

<template>
  <VPApi v-if="frontmatter.layout === 'api'">...</VPApi>
  <DefaultTheme.Layout v-else />
</template>
```

```js
// .vitepress/theme/index.js

import DefaultTheme from 'vitepress/theme'
import Layout from './Layout.vue'

export default {
  extends: DefaultTheme,
  Layout
}
```

**Comment by xsjcTony** [comment]
Agreed, but the current default theme's extensibility is not that ideal...
Also, I don't think they are clearly "internals", if I want to customize deeply it makes a lot of sense to use those "internal" composables to satisfy my needs, is it?

**Comment by xsjcTony** [comment]
I'd say expose other composables just like exposing `useSidebar` in `DefaultTheme` namespace is totally acceptable in my opinion.
https://vitepress.dev/reference/default-theme-sidebar#usesidebar

**Comment by xsjcTony** [comment]
@brc-dd I think... we should definitely expose all those internal types, including `composables` and `components`...

Since https://vitepress.dev/guide/extending-default-theme#overriding-internal-components is an absolutely valid use-case, and when overriding it's very likely to use some related internal components.

Giving you an example here, which is my real use case:
I'm trying to modify the content in the CARD of HOME FEATURES, which is `VPFeature.vue` internal component. However, since  I don't want to totally re-design the card, hence related internal components like `VPImage`, and `VPLink` will remain there for use.
But when I'm trying to run `vue-tsc`, I can't get it pass...

For now, I have to use either `// @ts-expect-error` or remove `vue-tsc` on `pnpm build`, but none of them make sense.

Just like I said, as long as `overriding internal components` is an official way to do the customization, then those types should NOT be considered as ***internal*** in my opinion. They should either being exported separately (which I recommend, since it's still the rare use case) or bundled together in the `DefaultTheme` namespace

`config.ts`
```ts
export default defineConfig({
  // vite
  vite: {
    resolve: {
      alias: [
        {
          find: /^.*\/VPFeature\.vue$/,
          replacement: resolvePath('./theme/components/internal/VPFeature.vue'),
        },
      ],
    },
  },
});
```
`MyVPFeature.vue`
```vue
<script setup lang="ts">
import VPIconArrowRight from 'vitepress/dist/client/theme-default/components/icons/VPIconArrowRight.vue';
import VPImage from 'vitepress/dist/client/theme-default/components/VPImage.vue';
import VPLink from 'vitepress/dist/client/theme-default/components/VPLink.vue';
import type { DefaultTheme } from 'vitepress/theme';
// ......
</script>
<!-- template and style... -->
```
![Snipaste_2023-07-18_17-42-33](https://github.com/vuejs/vitepress/assets/45784210/b9f0ed46-3727-43ac-97c5-ab650ab839a0)

**Comment by userquin** [comment]
@xsjcTony have you added vitepress to your tsconfig `compilerOptions.types`?

![imagen](https://github.com/vuejs/vitepress/assets/6311119/45ab2955-744e-4c1d-a7aa-053769efd445)

![imagen](https://github.com/vuejs/vitepress/assets/6311119/9e26c3db-20de-4a66-9725-7c0a4b90dc9d)

**Comment by xsjcTony** [comment]
@userquin It doesn't matter whether I add it into `compilerOptions.types`, because I don't have that options which means all types are visible by default. (Although I tried to define it, but no luck)

---

### For `useSidebar()` type
(applies no matter whether I use `Volar` or `TypeScript` in WebStorm)

1. I'm using `pnpm`, and haven't try `npm` yet. No idea if it's `pnpm` causing the issue.
2. When I can't go to type declaration by `ctrl + click` on `DocSideBar` , it just means there's some problem. The color of `DocSideBar` is also incorrect, it will be yellow if it's a TS type (can be parsed by IDE).
![Snipaste_2023-07-18_20-01-50](https://github.com/vuejs/vitepress/assets/45784210/b86b3353-e597-4846-9ded-af5f98dbe12d)
3. For actual usage, it's a little bit weird.
    - I can see the type here, but I can't click on it either. ![Snipaste_2023-07-18_20-05-31](https://github.com/vuejs/vitepress/assets/45784210/840e5c36-1964-46f5-92e8-7a29aef9a533)
    - Destructure won't work ![Snipaste_2023-07-18_20-06-01](https://github.com/vuejs/vitepress/assets/45784210/44cd21a3-62c9-4d83-ac42-b46809db97b4)
    - Property access works for simple type
     ![Snipaste_2023-07-18_20-06-45](https://github.com/vuejs/vitepress/assets/45784210/0fe53afb-4f12-4b3f-8d51-08a0c41a6e88) ![Snipaste_2023-07-18_20-07-18](https://github.com/vuejs/vitepress/assets/45784210/13d360ee-b2ca-4af0-aa53-56fd597a3524)
    - Property access on complex type sometimes work but sometimes doesn't
![Snipaste_2023-07-18_20-08-11](https://github.com/vuejs/vitepress/assets/45784210/8909ca3a-4e5f-401b-8951-f8c0442596af) ![Snipaste_2023-07-18_20-08-21](https://github.com/vuejs/vitepress/assets/45784210/35acdddc-393d-44be-8df0-5c17626dfce0)
    - Function doesn't work ![Snipaste_2023-07-18_20-08-29](https://github.com/vuejs/vitepress/assets/45784210/927e6c73-feca-47fc-8bdd-f696859940c6)
4. All the above only happens in IDE. There seems no issue when compiling using `vue-tsc`. I'm not sure whether because the type is correct or it's being evaluated into `any`

---

### For importing internal `.vue` components

It works in IDE (no error), but fails when checking by `vue-tsc`, as I posted above.

**Comment by xsjcTony** [comment]
@userquin Seems they are evaluated to `any` when running `vue-tsc`. That's definitely something wrong with exporting types
Note: results are the same no matter whether I include `vitepress` in `compilerOptions.types` or not.
![Snipaste_2023-07-18_20-20-11](https://github.com/vuejs/vitepress/assets/45784210/2b9c0ff4-6a1d-4cb4-b98a-e5cbf9d3072b)
![Snipaste_2023-07-18_20-20-25](https://github.com/vuejs/vitepress/assets/45784210/50ef9122-715a-4985-bc85-89d27d8b9928)

**Comment by xsjcTony** [comment]
@userquin

### For `useSidebar()` type

It's because of a typo in `theme.d.ts`. Here's the fix PR: https://github.com/vuejs/vitepress/pull/2643

### For importing internal `.vue` components

It works in IDE (no error), but fails when checking by `vue-tsc`, as I posted above.

It doesn't matter whether I add it into `compilerOptions.types`, because I don't have that options which means all types are visible by default. (Although I tried to define it, but no luck)

**Comment by eagereyes** [comment]
I recently posted [a related question in the discussion forum](https://github.com/vuejs/vitepress/discussions/2819). I'm creating a blog-type feed for the homepage on [my website](https://eagereyes.org/), and want to reuse as much of the default theme as possible. It would be great to be able to import `VPFeatures`, so I can pass in different props. Right now, I can only access `VPHomeFeatures`, so if I want them in more than one place, they have the same content since it's coming from the frontmatter. This actually works kind of okay for me for now (I use one block near the top and one at the bottom), but I'm trying to decide if it's worth copying a bunch of code just to get some nicer-looking links. It would be much easier if I could just access the underlying components.

So TL;DR is that I'd vote to make more default theme components available, especially the ones on the homepage.

**Comment by brc-dd** [comment]
fixed in #5396

however, the types don't mean the components are part of the public interface, we might still change the components between patch/minor versions. but mostly should be fine unless we remove some component or change the props, and also if you run vue-tsc as part of your build, it should catch up these changes now.
